### PR TITLE
Adding function to `join` multiple `TSFrame`s.

### PR DESCRIPTION
**Pull request by codetalker7**
This PR addresses issue https://github.com/xKDR/TSFrames.jl/issues/40. We have added a method for `join` which takes in a variable number of `TSFrame`s (this number should be at least 2). We have assumed that the `join`s are left associative. To keep the syntax similar to what we already have, we have added the join type as the keyword argument `jointype` (since the variable number of arguments needs to be the final positional argument). For more, please refer to the discussion in https://github.com/xKDR/TSFrames.jl/issues/40.

Corresponding documentation and tests have been added.

### CODE REVIEW

**Comment by chiraganand** [comment]
> This PR addresses issue #40. We have added a method for `join` which takes in a variable number of `TSFrame`s (this number should be at least 2). We have assumed that the `join`s are left associative. To keep the syntax similar to what we already have, we have added the join type as the keyword argument `jointype` (since the variable number of arguments needs to be the final positional argument). For more, please refer to the discussion in #40.
> 
> Corresponding documentation and tests have been added.

Yes, this sounds fine. Though, my personal preference would be to reuse the `on` keyword (instead of `jointype`) used by DataFrames.jl to maintain consistency in interfaces, should be easier for users.

**Comment by codetalker7** [comment]
> > This PR addresses issue #40. We have added a method for `join` which takes in a variable number of `TSFrame`s (this number should be at least 2). We have assumed that the `join`s are left associative. To keep the syntax similar to what we already have, we have added the join type as the keyword argument `jointype` (since the variable number of arguments needs to be the final positional argument). For more, please refer to the discussion in #40.
> > Corresponding documentation and tests have been added.
> 
> Yes, this sounds fine. Though, my personal preference would be to reuse the `on` keyword (instead of `jointype`) used by DataFrames.jl to maintain consistency in interfaces, should be easier for users.

Changed `jointype` to `on`.

**Comment by chiraganand** [comment]
> Changed `jointype` to `on`.

On second thoughts, `jointype` is a more appropriate name because `on` in DataFrames refers to a column name and not the kind of join, the latter being referred to by the method name itself.

Can you change it back to `jointype`? Sorry about this.

**Review by chiraganand — CHANGES_REQUESTED** [review]
The documentation also needs to be modified after this API change.

File: src/join.jl [hunk 1]
```diff
@@ -220,6 +232,62 @@ julia> join(ts1, ts2, JoinRight)
...
+ 2017-01-05  0.313244   0.313244   0.313244
+ 2017-01-06  0.662555   0.662555   0.662555
+ 2017-01-07  0.586022   0.586022   0.586022
+ 2017-01-08  0.0521332  0.0521332  0.0521332
+ 2017-01-09  0.26864    0.26864    0.26864
+ 2017-01-10  0.108871   0.108871   0.108871
+
 ```
 """
 function Base.join(ts1::TSFrame, ts2::TSFrame)
```
Comment (chiraganand) [inline src/join.jl:293]: The call to `join()` here needs to change to provide the keyword argument name.

File: src/join.jl [hunk 2]
```diff
@@ -247,5 +315,29 @@ function Base.join(ts1::TSFrame, ts2::TSFrame, ::Type{JoinRight})
     result = DataFrames.rightjoin(ts1.coredata, ts2.coredata, on=:Index, makeunique=true)
     return TSFrame(result)
 end
+
+function Base.join(
+    ts1::TSFrame,
+    ts2::TSFrame,
+    ts...;
+    jointype::T
```
Comment (chiraganand) [inline src/join.jl:323]: `jointype` should take a default value of `JoinAll`.

File: src/join.jl [hunk 3]
```diff
@@ -247,5 +315,29 @@ function Base.join(ts1::TSFrame, ts2::TSFrame, ::Type{JoinRight})
...
+        Type{JoinInner},
+        Type{JoinOuter},
+        Type{JoinLeft},
+        Type{JoinRight}
+    }
+}
+    if isempty(ts)
+        return Base.join(ts1, ts2, jointype)
+    else
+        return Base.join(Base.join(ts1, ts2, jointype), ts...; jointype=jointype)
```
Comment (chiraganand) [inline src/join.jl:338]: Will this work with. say, 5 ts objects because the methods above do not accept more than 2 arguments. I thought a better approach could be to make all the methods accept more than 2 args then this method here would directly call the other methods without modifying any arguments.

File: test/join.jl [hunk 4]
```diff
@@ -22,12 +37,28 @@ ts_outerjoin = join(ts1, ts2, JoinOuter)
 @test ts_outerjoin[1:Int(floor(DATA_SIZE/2)), :x1_1] == ts2[1:Int(floor(DATA_SIZE/2)), :x1]
 @test isequal(Vector{Missing}(ts_outerjoin[Int(floor(DATA_SIZE/2)) + 1:DATA_SIZE, :x1_1]), fill(missing, DATA_SIZE - Int(floor(DATA_SIZE/2))))
 
-ts_joinboth = join(ts1, ts2, JoinAll)
-@test propertynames(ts_joinboth.coredata) == [:Index, :x1, :x1_1]
-@test ts_joinboth[:, :Index] == ts1[1:DATA_SIZE, :Index]
-@test ts_joinboth[:, :x1] == ts1[1:DATA_SIZE, :x1]
-@test ts_joinboth[1:Int(floor(DATA_SIZE/2)), :x1_1] == ts2[1:Int(floor(DATA_SIZE/2)), :x1]
-@test isequal(Vector{Missing}(ts_joinboth[Int(floor(DATA_SIZE/2)) + 1:DATA_SIZE, :x1_1]), fill(missing, DATA_SIZE - Int(floor(DATA_SIZE/2))))
+ts_joinall = join(ts1, ts2, JoinAll)
```
Comment (chiraganand) [inline test/join.jl:40]: Can you add another test(s) for testing 4 or 5 ts objects join as a boundary test case as there are already 3 arguments in the method signature so the code might just work for 3?

**Review by chiraganand — CHANGES_REQUESTED** [review]

File: src/join.jl (hunk 3 above)
Comment (codetalker7) [reply src/join.jl:338]: Yes, this will work for 5 objects. For example, the following code works.

```julia
using TSFrames, DataFrames, Dates, Random, Statistics
dates1 = collect(Date(2017,1,1):Day(1):Date(2017,1,10))
dates2 = collect(Date(2017,1,1):Day(1):Date(2017,1,30))
ts1 = TSFrame(rand(length(dates1)), dates1)
ts2 = TSFrame(rand(length(dates2)), dates2)
ts3 = TSFrame(rand(length(dates2)), dates2)
ts4 = TSFrame(rand(length(dates2)), dates2)
ts5 = TSFrame(rand(length(dates1)), dates1)

ts_all = join(ts1, ts2, ts3, ts4, ts5)
ts_left = join(ts1, ts2, ts3, ts4, ts5; jointype=JoinLeft)
```
In this method we are just doing the following: if there are exactly 2 ts objects, just call one of the older methods. If there are more than two objects, then compute recursively. This is also how DataFrames.jl implements multiple joins.

We can modify the old methods to have multiple ts objects, but that will just be more code in my opinion. Here we can handle everything with one function.

File: test/join.jl (hunk 4 above)
Comment (codetalker7) [reply test/join.jl:60]: Added tests for 5 objects.

File: src/join.jl (hunk 3 above)
Comment (chiraganand) [reply src/join.jl:333]: > We can modify the old methods to have multiple ts objects, but that will just be more code in my opinion. Here we can handle everything with one function.

Wouldn't is be easier to maintain one type of logic in all methods rather than having multiple logics in multiple methods?

Comment (chiraganand) [reply src/join.jl:333]: Also, the question to ask is whether to even keep the old methods which dispatch on join types? Shouldn't a single method even if it is recursive work? Do we even need the `JoinXXX` types now?

Comment (chiraganand) [reply src/join.jl:333]: > ... This is also how DataFrames.jl implements multiple joins.

`DataFrames.join*()` functions are recursive then wouldn't `TSFrames.join()` call add another layer of recursive calls on top? Shouldn't the extra layer increase the time complexity by a factor of `log(n)`? In that case, I feel we should collect all the arguments together (`ts1, ts2, ts3, ...`) and then call `DataFrames.join()` in one go.

**Comment by codetalker7** [comment]
The latest commit replaces everything by a single function. Note that here we are using a `for` loop; we had to do this because some `DataFrames` join function like `leftjoin` and `rightjoin` work on *only* two `DataFrame` objects (and not more than two objects).

**Comment by chiraganand** [comment]
> The latest commit replaces everything by a single function. Note that here we are using a `for` loop; we had to do this because some `DataFrames` join function like `leftjoin` and `rightjoin` work on _only_ two `DataFrame` objects (and not more than two objects).

This should be fine as long as there are test cases to make sure the code is working fine for `leftjoin` et al for more than 2 or 3 `TSFrame` objects.

**Comment by codetalker7** [comment]
> > The latest commit replaces everything by a single function. Note that here we are using a `for` loop; we had to do this because some `DataFrames` join function like `leftjoin` and `rightjoin` work on _only_ two `DataFrame` objects (and not more than two objects).
> 
> This should be fine as long as there are test cases to make sure the code is working fine for `leftjoin` et al for more than 2 or 3 `TSFrame` objects.

Hi. Yes, we have tests for 2, 3 and 5 `TSFrame` objects for each function.

**Review by chiraganand — APPROVED** [review]
